File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys

input = sys.stdin.readline
n = int(input())

colorArr = [list(map(int, input().split(" "))) for _ in range(n)]

def solution():
    dp = [[0]*3 for _ in range(n)];
    
    dp[0][0] = colorArr[0][0];
    dp[0][1] = colorArr[0][1];
    dp[0][2] = colorArr[0][2];
    
    for i in range(1,n) :
        dp[i][0] = min(dp[i-1][1],dp[i-1][2]) + colorArr[i][0];
        dp[i][1] = min(dp[i-1][0],dp[i-1][2]) + colorArr[i][1];
        dp[i][2] = min(dp[i-1][0],dp[i-1][1]) + colorArr[i][2];
    
    return min(dp[n-1])
    
print(solution());

File: 백준/Silver/1149. RGB거리/RGB거리.js
//백준 제출 시
const fs = require("fs");
const input = fs.readFileSync(0, "utf-8").trim().split("\n");

const n = Number(input[0]);
const colorArr = input.slice(1).map((i) => i.split(" ").map(Number));

function solution() {
  const dp = Array.from({ length: n }, () => Array(3).fill(0));
  

  //작은 문제로 큰문제를 푼다.
  dp[0][0] = colorArr[0][0];
  dp[0][1] = colorArr[0][1];
  dp[0][2] = colorArr[0][2];

  for (let i = 1; i < n; i++) {
    dp[i][0] = Math.min(dp[i - 1][1], dp[i - 1][2]) + colorArr[i][0];
    dp[i][1] = Math.min(dp[i - 1][0], dp[i - 1][2]) + colorArr[i][1];
    dp[i][2] = Math.min(dp[i - 1][0], dp[i - 1][1]) + colorArr[i][2];
  }

  return Math.min(...dp[n - 1]);
}

console.log(solution());
